[BLOG] Add "Writing fast string ufuncs for NumPy 2.0" (#839)
Quansight/Quansight-website: apps/labs/posts/numpy-string-ufuncs.md

string dtypes, one for Unicode (UTF-32) and one for bytes (ASCII). Both of these
dtypes are fixed-length dtypes, which means that the maximum length of each array
item is specified by the dtype. Let's take the following example:

```python3
>>> import numpy as np
>>> s = np.array(["abcd"])
>>> s
array(['abcd'], dtype='<U4')
```

The `dtype='<U4'` tells us that the strings in this array have a maximum length
of four characters. Were we to assign an item of the array to a string that's
longer than that, the excess characters would be stripped:

```python3
>>> s[0] = 'efghi'
>>> s
array(['efgh'], dtype='<U4')
```

That doesn't mean, however, that all of the array items have to be exactly four
characters long. We can have array items that are smaller than that. The underlying
array buffer still allocates the number of bytes corresponding to four characters
(4 bytes for ASCII, 16 bytes for UTF-32), but the rest of the bytes are zero. That
has the side-effect that both of these dtypes strip all trailing zero bytes.

```python3
>>> s = np.array(["abcd", "ef"])
>>> s
array(['abcd', 'ef'], dtype='<U4')
>>> s[0] = "h\0\0"
>>> s

> array(['h', 'ef'], dtype='<U4')

Reviewer: you can illustrate how the actual array is laid out in memory printing the result of `s.tobytes()`, I sometimes find that useful to understand this.

easy enough for NumPy to do, since booleans are fixed-size dtypes. However, string
dtypes, both unicode and bytes, are parametric dtypes, meaning that their size
is dependent on the array items, specifically their lengths. In order for NumPy
to be able to register the loop, it needs a `resolve_descriptors` function which
specifies the parameters for all the necessary dtypes. This function has the
following signature:

```cpp
NPY_CASTING resolve_descriptors(struct PyArrayMethodObject_tag *method,
                                PyArray_DTypeMeta **dtypes,
                                PyArray_Descr **given_descrs,
                                PyArray_Descr **loop_descrs,
                                npy_intp *view_offset)
```

The important arguments here are `given_descrs` and `loop_descrs`:
- `given_descrs` is an array of input and output (if any were given
   by the user) descriptor instances. A descriptor instance is a descriptor
   for an array, and includes information about the size of each item.
- `loop_descrs` is an array of descriptors that must be filled in by the resolve
   descriptors implementation. These will be the descriptors that will be used

>    in order to call the loop with the correct information.

Reviewer: maybe also link to the numpy 2.0 ArrayMethod docs which go over all this

Reviewer: https://numpy.org/devdocs/reference/c-api/array.html#c.NPY_METH_resolve_descriptors (not yet available in a build of the docs for a stable release)

`isalpha` though, they both accept four input arguments: the buffer to search in,
the substring to search for, and starting and ending indices (if one wants to
specify in what part of the buffer to search). The starting and ending indices
can be of any integer dtype (int8, int16, int32 etc.), but we only want to write
one loop for all of them. To achieve that we need a promoter.

A promoter is a function that specifies the promotion rules for a ufunc. In general,
a ufunc will search for a loop that matches the input dtypes exactly. In order
to be able to accept more than one input dtype, and have all of those dispatch
to the same loop, we should register a promoter. A promoter has the following
signature:

```cpp
int promoter(PyUFuncObject *ufunc,
             PyArray_DTypeMeta **op_dtypes,
             PyArray_DTypeMeta **signature,
             PyArray_DTypeMeta **new_op_dtypes)
```

The two important arguments here are `op_dtypes` and `new_op_dtypes`. Similar
to the `resolve_descriptors` function, `op_dtypes` are the dtypes the user passed
into the ufunc. `new_op_dtypes` is an array of dtypes that must be filled with

> the dtypes the input should be promoted to before searching for a loop.

Reviewer: This is better documentation than what is currently in the numpy docs! Consider improving those docs if you have the time and motivation. https://numpy.org/devdocs/reference/c-api/array.html#c.PyArrayMethod_PromoterFunction

Author: Will do!

`replace` accepts four arguments: the buffer to search, the substring to search for
and replace, the replacement string, and a final integer argument that specifies
the maximum number of replacements to do. Putting everything we've covered so far
together gives us the following tasks:

1. Write a loop for the string dtypes and int64.
2. Add a promoter that promotes other integer dtypes to int64 for the last argument.
3. Write a `resolve_descriptors` function that specifies the output dtype size.

But, wait! Those that paid attention to the section about `add` will remember that
the `resolve_descriptors` function does not get access to the data itself; it only
gets access to the array descriptors that contain meta information about the array.
By just looking at the input dtype sizes though, we do not know how many replacements
we'll have to do, and how long the final string will be. In order to circumvent
that, we have to create a Python wrapper around the ufunc written in C++. But,
before we explain how the wrapper works, let's talk about another feature of ufuncs:
the `out` optional keyword argument.

> If the optional `out` keyword argument is `None`, which is the default, a new array

Reviewer: Maybe worth describing what `out` does earlier in the blog where you go over `add` above, since this is a more fundamental concept that readers who are familiar with numpy might know about, but probably don't know how it's handled in a ufunc implementation.

Author: Moved to the `add` section. Maybe worth checking that it doesn't look out of place there.

    if (loop_descrs[1] == NULL) {
        return _NPY_ERROR_OCCURRED_IN_CAST;
    }

    // Create a new descriptor instance from the first input and add
    // the element size of the second input.
    loop_descrs[2] = PyArray_DescrNew(loop_descrs[0]);
    if (loop_descrs[2] == NULL) {
        return _NPY_ERROR_OCCURRED_IN_CAST;
    }
    loop_descrs[2]->elsize += loop_descrs[1]->elsize;

    return NPY_NO_CASTING;
}
```

It does the following:
- Specifies that the input dtypes are the ones that should be
  used for passing information to the loop. In doing so, it also ensures that
  the dtypes are in canonical representation, meaning in the underlying system's
  native byte order.
- For the output dtype (`loop_descrs[2]`), we create a new descriptor instance
  from one of the inputs (either one would be fine), and then add the other's
  element size. This way, we get an output dtype with a size that's equal to
  the sum of the sizes of the two input dtypes.

After registering the `resolve_descriptors` function, everything works correctly
and `np.add` now works for string dtypes as well!

## `find` / `rfind`: Nothing new here, right?

> Next in our list of functions to implement as ufuncs was `find` and `rfind`. Like

Reviewer 2: ```suggestion
Next in our list of functions to implement as ufuncs were `find` and `rfind`. Like
```

the maximum number of replacements to do. Putting everything we've covered so far
together gives us the following tasks:

1. Write a loop for the string dtypes and int64.
2. Add a promoter that promotes other integer dtypes to int64 for the last argument.
3. Write a `resolve_descriptors` function that specifies the output dtype size.

But, wait! Those that paid attention to the section about `add` will remember that
the `resolve_descriptors` function does not get access to the data itself; it only
gets access to the array descriptors that contain meta information about the array.
By just looking at the input dtype sizes though, we do not know how many replacements
we'll have to do, and how long the final string will be. In order to circumvent
that, we have to create a Python wrapper around the ufunc written in C++ and use
the `out` kwarg, which I described above, to pass an array with the
correct dtype into the `replace` ufunc. All we have to do is find out what the maximum
length is, which, through some NumPy magic, is easy enough:

```python3
buffersizes = str_len(arr) + counts * (str_len(new) - str_len(old))
```

> It's just the diffence in size between the substring to be replaced and the replacement

Reviewer 2: ```suggestion
It's just the difference in size between the substring to be replaced and the replacement
```

string, multiplied by the number of replacements. This leads to to the Python
wrapper looking something like this:

> ```python3

Reviewer 2: ```suggestion
```python
```

def replace(a, old, new, count=-1):
    # Convert all input arguments to arrays
    arr = np.asanyarray(a)
    a_dt = arr.dtype
    old = np.asanyarray(old, dtype=getattr(old, 'dtype', a_dt))
    new = np.asanyarray(new, dtype=getattr(new, 'dtype', a_dt))

>     # Count the number of occurences of `old` in `a`

Reviewer 2: ```suggestion
    # Count the number of occurrences of `old` in `a`
```

    # and compare that to the number of maximum replacements
    max_int64 = np.iinfo(np.int64).max
    counts = _count_ufunc(arr, old, 0, max_int64)
    count = np.asanyarray(count)
    counts = np.where(count < 0, counts, np.minimum(counts, count))

>     # Compute buffersizes, construct output array and call ufunc

Reviewer 2: ```suggestion
    # Compute buffer sizes, construct output array and call ufunc
```

`resolve_descriptors` function:

```cpp
NPY_CASTING
string_replace_resolve_descriptors(PyArrayMethodObject *self,
                                   PyArray_DTypeMeta *dtypes[5],
                                   PyArray_Descr *given_descrs[5],
                                   PyArray_Descr *loop_descrs[5],
                                   npy_intp *view_offset)
{
    // Error in case the last item of given_descrs (the output descriptor)
    // is NULL which means that out=None.
    if (given_descrs[4] == NULL) {
        PyErr_SetString(PyExc_ValueError, "out kwarg should be given");
        return _NPY_ERROR_OCCURRED_IN_CAST;
    }

    loop_descrs[0] = NPY_DT_CALL_ensure_canonical(given_descrs[0]);
    if (loop_descrs[0] == NULL) {
        return _NPY_ERROR_OCCURRED_IN_CAST;
    }

    // Omit the checks for NULL for conciseness (they're there in reality)
    loop_descrs[1] = NPY_DT_CALL_ensure_canonical(given_descrs[1]);
    loop_descrs[2] = NPY_DT_CALL_ensure_canonical(given_descrs[2]);
    loop_descrs[3] = NPY_DT_CALL_ensure_canonical(given_descrs[3]);
    loop_descrs[4] = NPY_DT_CALL_ensure_canonical(given_descrs[4]);

    return NPY_NO_CASTING;
}
```

> With this function in place the `_replace` internal ufunc will fail as expected when

Reviewer 2: ```suggestion
With this function in place, the `_replace` internal ufunc will fail as expected when
```

`out` is not provided. To get a better interface and avoid the need for an `out` array
one can use the `numpy.strings.replace` wrapper.

> ```python3

Reviewer 2: ```suggestion
```python
```

>>> a = np.array(["abc"])
>>> out = np.empty_like(a, shape=a.shape, dtype="U3")
>>> _replace(a, "b", "c", 1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: out kwarg should be given
>>> _replace(a, "b", "c", 1, out=out)
array(['acc'], dtype='<U3')
>>> np.strings.replace(a, "b", "c")
array(['acc'], dtype='<U3')
```

## Results

> The benchmarks we ran were not too extensive, but the little results we had

Reviewer 2: ```suggestion
The benchmarks we ran were not too extensive, but the few results we had
```

were already so promising, that we decided that doing all of this for all of the
string operations was definitely going to be worth it.

> More specifically, we ran two experiments. One with a very small list with

Reviewer 2: ```suggestion
More specifically, we ran two experiments. One with a very small array with
```

> relatively big strings and one with somewhat bigger list that only contained

Reviewer 2: ```suggestion
relatively long strings and one with somewhat larger array that only contained
```

> small strings. Here's an illustration of the exact results:

Reviewer 2: ```suggestion
short strings. Here's an illustration of the exact results:
```

<img
  src='/posts/numpy-string-ufuncs/benchmarks.png'
  alt="A bar plot showing the performance increase between the old approach and the new ufuncs across two different benchmarks" />

The results clearly illustrate a big performance improvement, especially for

> bigger arrays. In the benchmarks we ran with a 1000-element array, there was

Reviewer 2: ```suggestion
larger arrays. In the benchmarks we ran with a 1000-element array, there was
```

a performance increase of 150x-492x depending on the string operation, while

> there was a much moderate speed-up of 4x-11x for smaller two-element arrays.

Reviewer 2: ```suggestion
there was a more moderate speed-up of 4x-11x for smaller two-element arrays.
```

This goes to show that operating on the raw C data, instead of doing the CPython
PyObject dance explained above, has a very significant effect.

## Conclusion

> In this blog post, we explained how an observation from our client, D.E. Shaw,

Reviewer 2: ```suggestion
In this blog post, we explained how an observation from our client, the D.E. Shaw group,
```

led to a complete rework of how string operations are done for NumPy arrays.
In the process, we went over how to write ufuncs that operate on the raw C
data buffer of the array and we described how the `resolve_descriptors` function
can be used to let NumPy know what the output dtype will be for parametric
dtypes. We also briefly touched upon how to write a promoter function so that
we can write only one loop that can operate on multiple dtypes.

But that's not all! While all of this was brewing, [Nathan Goldbaum](https://github.com/ngoldbaum)
was also working on a variable length string dtype with support for UTF-8.
Combining these two work streams, NumPy 2.0 comes with a new `numpy.strings`

> namepsace that implements fast ufuncs for most of the string operations (with more

Reviewer 2: ```suggestion
namespace that implements fast ufuncs for most of the string operations (with more
```

to come soon!) with support for byte strings (ASCII) and unicode (both UTF-8 and
UTF-32).

## Acknowledgements

Thanks to [Nathan Goldbaum](https://github.com/ngoldbaum), [Sebastian Berg](https://github.com/seberg),
[Marten van Kerkwijk](https://github.com/mhvk) and [Matti Picus](https://github.com/mattip) for their
countless reviews and all of their patience.

> Big thanks to the D.E. Shaw group for sponsoring this work.

Reviewer 2: ```suggestion
A big thank you to the D.E. Shaw group for sponsoring this work.
```
